---
title: "scRNAseq_ASXLaligned"
author: "Kate Randall"
date: "2025-05-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages


```{r}
library(plyr)
library(tidyverse)
library(ggplot2)
library(ggpackets)
library(patchwork)
library(glmGamPoi)
library(presto)
library(BPCells)
library(RANN)
library(Rtsne)
library(irlba)
library(png)
library(qqconf)
library(Matrix)
library(lmtest)
library(metap)
library(multtest)
library(mime)
library(ellipsis)
library(lazyeval)
library(stringi)
library(reshape2)
library(polyclip)
library(goftest)
library(RColorBrewer)
library(viridis)
library(beepr)
library(SeuratObject)
library(Seurat)
#library(ggseurat)
library(ggpackets)
library(sctransform)
library(scCustomize)

# set seed for reproducibility
set.seed(123456)

# set assay version to Seurat v3
#options(Seurat.object.assay.version = "v3")

```


# Import scRNA-seq dataset


```{Import scRNA-seq files (kallisto/bustools output)}

# Define path to kallisto_out
base_dir <- "/media/bret/Data2/align_to_ASXL_transcriptome/kallisto_out"

samples <- c(
  "0_hrs",
  "3_hrs",
  "6_hrs",
  "12_hrs",
  "18_hrs",
  "24_hrs",
  "36_hrs"
)

output_dir <- "seurat_objects"
dir.create(output_dir, showWarnings = FALSE)

for (sample in samples) {
  sample_dir <- file.path(base_dir, sample)
  mtx <- readMM(file.path(sample_dir, "genes.mtx"))
  barcodes <- readLines(file.path(sample_dir, "genes.barcodes.txt"))
  genes <- readLines(file.path(sample_dir, "genes.genes.txt"))
  
  colnames(mtx) <- barcodes
  rownames(mtx) <- genes
  
  seurat_obj <- CreateSeuratObject(counts = mtx, project = sample)
  seurat_obj$timepoint <- sample
}


# SAVE & LOAD
save(hr0, hr3, hr6, hr12, hr18, hr24, hr36, file="hr.meta.RData")
#load(file = "hr.meta.RData")

```



# Quality Control Metrics



## Merge timepoints (for QC only)

```{Merge timepoints - QUICKLY with multicore!}

# in Terminal window!

R

# load packages and inputs as above
library(future)
library(Seurat)

load(file = "hr.meta.RData")


# REMOVE LATER force v3 assay (instead of v5)
#options(Seurat.object.assay.version = "v3")

# set up for parallelization 
plan("multicore", workers = parallel::detectCores())
options(future.globals.maxSize = 100 * 1024^3) # use <100 GB RAM

# log compute time
start_time <- Sys.time()

# create merged object (multicore)
hr.merged <- value(future({
  merge(x = hr0,
        y = list(hr3, hr6, hr12, hr18, hr24, hr36),
        add.cell.ids = c("0hrs", "3hrs", "6hrs", "12hrs", "18hrs", "24hrs", "36hrs"))
}))

save(hr.merged, file = "hr.mergedASXL.RData")

# report compute time
end_time <- Sys.time()
cat("Merge compute time:", end_time - start_time)

```

## View raw counts (before any filtering or normalization)

```{r}
# load back into RStudio
load(file = "hr.mergedASXL.RData")

# clean up env
rm(hr0, hr3, hr6, hr12, hr18, hr24, hr36)


raw.metadata <- hr.merged@meta.data
head(raw.metadata)
unique(raw.metadata$timepoint)

#force chronological timepoint order (default is alphanumeric)
raw.metadata$timepoint <- factor(raw.metadata$timepoint,
                             levels = c("hr0", "hr3", "hr6", "hr12", "hr18", "hr24", "hr36"))


# set custom ggplot theme
theme <- ggpacket() +
  theme(text=element_text(size=18, color="black"),
        rect=(element_blank()),
        axis.line=element_line(),
        axis.text=element_text(size=14, color="black"),
        plot.title=element_text(hjust=0.5, face = "bold", size = 22))


# Plot (bar) nCells per timepoint (before normalization, scaling, or filtration)
nCells_raw <- raw.metadata %>%
  ggplot(aes(x= timepoint, fill= timepoint)) +
  geom_bar() +
 # theme() +
  scale_y_continuous(expand=(c(0,0))) +
  scale_fill_viridis_d() +
  ggtitle("nCells (raw counts)")
nCells_raw
ggsave("nCells_raw", plot = nCells_raw, device = "pdf")

```



```{Data wrangling}

# reload back in RStudio
load(file = "hr.mergedASXL.RData")

# copy cell ids into new column "timepoint"
hr.merged[["timepoint"]] <- sapply(Cells(hr.merged), function(x) strsplit(x, "_")[[1]][[1]])
unique(hr.merged@meta.data$timepoint)

# REMOVE LATER other metadata wrangling
# change hr.merged from v5 to V3
#hr.merged[["RNA"]] <- as(object = hr.merged[["RNA"]], Class = "Assay")
# pull metadata for custom plotting
#metadata <- hr.merged@meta.data
#unique(sapply(X = strsplit(colnames(metadata), split = "_"), FUN = "[", 1))
#head(metadata)
# add cells column
#hr.merged_metadata$cells <- rownames(hr.merged_metadata)


# SAVE & LOAD          
save(hr.merged, file = "hr.mergedASXL.RData")
#load(file = "hr.mergedASXL.RData")


```


## Remove empty droplets

```{r}
# first, just filter out low nCount_RNA "cells"
hr.merged <- subset(hr.merged, subset = nCount_RNA > 200)

# run Cellbender (see python scripts) and import results
# ERROR: cant pickle weakrefs 

# compare Cellbender results to nCount_RNA filter
# would go here if I could get Cellbender to work

```

